packages/datadog-plugin-aws-sdk/src/index.js

```diff
@@ -0,0 +1,109 @@
+'use strict'
+
+const Tags = require('opentracing').Tags
+const analyticsSampler = require('../../dd-trace/src/analytics_sampler')
+const awsHelpers = require('./helpers')
+
+function createWrapRequest (tracer, config) {
+  config = normalizeConfig(config)
+
+  return function wrapRequest (request) {
+    return function requestWithTrace (operation, params, cb) {
+      const serviceName = awsHelpers.normalizeServiceName(this)
+      const childOf = tracer.scope().active()
+      const baseTags = {
+        [Tags.SPAN_KIND]: 'client',
+        'span.type': 'http',
+        'service.name': config.service ? `${config.service}-${serviceName}` : `${tracer._service}-${serviceName}`,
+        'aws.agent': 'js-aws-sdk',
+        'aws.operation': operation,
+        'aws.region': this.config.region,
+        'aws.service': serviceName,
+        'aws.url': this.endpoint && this.endpoint.href,
+        'component': serviceName
+      }
+      let span
+
+      if (typeof cb === 'function') {
+        span = tracer.startSpan('aws.http', {
+          childOf,
+          tags: baseTags
+        })
+
+        analyticsSampler.sample(span, config.analytics)
+
+        // when passed a callback makeRequest will call `.send` on the AWS.Request
+        // before we have a chance to add our listeners, so we have to wrap
+        // the callback in order know when the span has finished
+        // https://github.com/aws/aws-sdk-js/blob/
+        // 9c191bbdbf32a8a3fa31219e369006f852318a1f/lib/service.js#L202-L206
+        return tracer.scope().activate(span, () => {
+          return request.call(
+            this,
+            operation,
+            params,
+            awsHelpers.wrapCallback(tracer, span, cb, childOf, config, serviceName)
+          )
+        })
+      } else {
+        const awsReq = request.apply(this, arguments)
+        const boundAwsReq = tracer.scope().bind(awsReq, childOf)
+
+        // The state machine has garaunteed first and last events, validate / complete
+        // instrumenting those ensures requests w/out callback gets started/ended correctly
+        // https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L142
+        // https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L328-L332
+        boundAwsReq.on('validate', () => {
```

Comment:
Can this fail? How is the failure reported?

Reply:
see: https://github.com/DataDog/dd-trace-js/pull/856#discussion_r382562374